### PR TITLE
[MM-17388] Selecting an emoji from emoji picker should insert it at the cursor position

### PR DESCRIPTION
**Pull request by AGMETEOR**
**Summary**

MM-17388- Selecting an emoji from emoji picker should insert it at the cursor position

**Ticket Link**
[https://mattermost.atlassian.net/browse/MM-17388](https://mattermost.atlassian.net/browse/MM-17388)
Fixes https://github.com/mattermost/mattermost-server/issues/11954

### CODE REVIEW

**Comment by hanzei** [comment]
Hey @AGMETEOR,

just want to check if we can help in any way to get this PR finished.

**Comment by AGMETEOR** [comment]
Hey @hanzei , sorry for the delay. I got issues with my laptop and wasn't able to work... I have fixed that now and will deliver this mid week.

**Comment by hanzei** [comment]
Hey @AGMETEOR, thanks for the update. :+1:  No hurries, just wanted to make sure, this doesn't get forgotten.

**Comment by AGMETEOR** [comment]
Hi @hanzei this is ready for review but some steps in the CI are blocked.

**Comment by AGMETEOR** [comment]
> @AGMETEOR this is sweet. One small thing,
> 
> 1. When an emoji is inserted it moves the cursor to the end of the input. I would expect that after an emoji is inserted, the cursor remains after the space following the emoji (ie before the comma in this example below).
> 
> emoji

@esethna thank you for pointing that out. I had totally missed it. Let me work on that improvement.

**Comment by AGMETEOR** [comment]
@esethna I have implemented the feedback and also took care of the edge cases I could think of. Looking forward to more feedback if you have some.

**Comment by esethna** [comment]
Thanks @AGMETEOR!

cc// @hanzei if you can help get the test server going or look into the pipeline issues

**Comment by AGMETEOR** [comment]
Hello @hanzei and @esethna , just bubbling this up again... wondering if you have any more feedback on this PR.

File: components/create_comment/create_comment.jsx [hunk 1]
```diff
@@ -548,7 +559,29 @@ export default class CreateComment extends React.PureComponent {
         this.draftsForPost[this.props.rootId] = updatedDraft;
     }
 
+    handleMouseUp = (e) => {
+        e.persist();
+        const caretPosition = Utils.getCaretPosition(e.target);
+        const target = e;
```
Comment (hmhealey) [inline components/create_comment/create_comment.jsx]: Shouldn't this be `const target = e.target;`? That might get rid of the need to persist the event as well

File: components/create_comment/create_comment.jsx [hunk 2]
```diff
@@ -344,16 +347,24 @@ export default class CreateComment extends React.PureComponent {
...
         } else {
-            newMessage = `${draft.message} :${emojiAlias}: `;
+            const {draft: {message}} = this.state;
+            const {firstPiece, lastPiece} = splitMessageBasedOnCaretPosition.call(this, message);
+
+            // check whether the first piece of the message is empty when cursor is placed at beginning of message and avoid adding an empty string at the beginning of the message
+            newMessage = firstPiece === '' ? `:${emojiAlias}: ${lastPiece} ` : `${firstPiece} :${emojiAlias}: ${lastPiece} `;
+
+            const newCaretPosition = `${firstPiece} :${emojiAlias}: `.length;
+            repositionCaret(target, newCaretPosition);
```
Comment (hmhealey) [inline components/create_comment/create_comment.jsx]: Instead of using a helper function that runs with `requestAnimationFrame`, maybe it would be cleaner to use the second argument of `setState` to update the actual caret position after the state has been updated?
```javascript
this.setState({
    caretPosition: newCaretPosition,
}, () => {
    Utils.setCaretPosition(textbox, newCaretPosition);
});
```

File: components/create_comment/create_comment.jsx [hunk 3]
```diff
@@ -548,7 +559,29 @@ export default class CreateComment extends React.PureComponent {
...
+        this.setState({
+            caretPosition,
+            target,
+        });
+    }
+
     handleKeyDown = (e) => {
+        e.persist();
+
+        // A bit of a hack. Better to use keyup instead.
```
Comment (hmhealey) [inline components/create_comment/create_comment.jsx]: Would it be difficult to add a keyup handler for this? Using `setTimeout` like this seems like it coudl cause other issues later on

File: components/create_comment/create_comment.jsx [hunk 4]
```diff
@@ -548,7 +559,29 @@ export default class CreateComment extends React.PureComponent {
...
+        // We therefore use setTimeout to wait for the right value of caretPosition
+        // and then update state.
+
+        setTimeout(() => {
+            const caretPosition = Utils.getCaretPosition(e.target);
+            const target = e;
+            this.setState({caretPosition, target});
```
Comment (hmhealey) [inline components/create_comment/create_comment.jsx]: Instead of storing the target in the state, what if we used a ref to access the inner textbox instead? The `Textbox` component already accesses the inner textbox in places [like this](https://github.com/mattermost/mattermost-webapp/blob/master/components/textbox/textbox.jsx#L125), so I don't know if it would be hard to expose that to the `CreateComment` and `CreatePost` components. State doesn't seem like the best place to store that reference since it's not used for rendering the component

File: package.json [hunk 5]
```diff
@@ -96,7 +96,7 @@
     "babel-eslint": "10.0.2",
     "babel-jest": "24.8.0",
     "babel-loader": "8.0.6",
-    "babel-plugin-typescript-to-proptypes": "0.17.1",
+    "babel-plugin-typescript-to-proptypes": "^0.17.1",
```
[the remark this answers is not in the document]
Comment (AGMETEOR) [reply package.json]: @hmhealey made the revert.

File: utils/post_utils.jsx [hunk 6]
```diff
@@ -449,3 +449,24 @@ export function createAriaLabelForPost(post, author, isFlagged, reactions, intl)
 
     return ariaLabel;
 }
+
+// Splits text message based on the current caret position
+export function splitMessageBasedOnCaretPosition(message) {
+    if (this.state) {
```
[the remark this answers is not in the document]
Comment (AGMETEOR) [reply utils/post_utils.jsx]: @hmhealey I have made the suggested changes on this :)

File: components/create_comment/create_comment.jsx (hunk 1 above)
Comment (AGMETEOR) [reply components/create_comment/create_comment.jsx]: @hmhealey it's actually me storing the event. I should have named it appropriately, something like `targetEvent`. But after implementing your last feedback on this PR, that's now deleted.

File: components/create_comment/create_comment.jsx (hunk 2 above)
Comment (AGMETEOR) [reply components/create_comment/create_comment.jsx]: @hmhealey I have made the suggested changes on this. Works fine, thanks! :)

File: components/create_comment/create_comment.jsx (hunk 3 above)
Comment (AGMETEOR) [reply components/create_comment/create_comment.jsx]: @hmhealey I actually wanted to create one but thought to reuse an already existing event handler was appropriate. I have now added a keyup handler. Thanks for the review!

File: components/create_comment/create_comment.jsx (hunk 4 above)
Comment (AGMETEOR) [reply components/create_comment/create_comment.jsx]: @hmhealey implemented the suggested changes on this and picked a `ref` from the inner textbox

**Comment by AGMETEOR** [comment]
@hmhealey and @hanzei changed files have increased because I realized the issue also affected the `edit_post_modal` and I made fixes in that file too :)

**Comment by hanzei** [comment]
Gentle reminder to review this @sudheerDev

File: components/create_comment/create_comment.jsx [hunk 7]
```diff
@@ -549,6 +563,20 @@ export default class CreateComment extends React.PureComponent {
...
+        });
+    }
+
+    handleKeyUp = (e) => {
+        const caretPosition = Utils.getCaretPosition(e.target);
+        this.setState({
+            caretPosition,
+        });
+    }
+
```
Comment (sudheerDev) [inline components/create_comment/create_comment.jsx]: One minor nit here: We can change this to be one func as `handleKeyEvent` both are functionally same. i am 0/5 on the function name here.

Comment (AGMETEOR) [reply components/create_comment/create_comment.jsx]: @sudheerDev condensed the two functions into one. Thanks for the review :)

**Comment by sudheerDev** [comment]
@AGMETEOR Thanks for the contribution 👍
